src/Catalog/Helpers/Db2CatalogSqlQueries.cs

```diff
@@ -0,0 +1,122 @@
+        }
+
+        public async Task<IReadOnlyCollection<DateTime>> GetTimestampsAsync(
+            SqlConnection sqlConnection,
+            SqlTransaction sqlTransaction,
+            int top,
+            DateTime since,
+            string columnName)
+        {
+            var nextTimestampsQuery = $@"
+                        SELECT TOP {top} P.[{columnName}]
+                        FROM [dbo].[Packages] AS P
+                        WHERE [{columnName}] IS NOT NULL AND [{columnName}] > @{CursorParameterName}
+                        GROUP BY [{columnName}]
+                        ORDER BY [{columnName}]";
+
+            var nextTimestampsCommand = new SqlCommand(nextTimestampsQuery, sqlConnection, sqlTransaction)
+            {
+                CommandTimeout = _commandTimeout
+            };
+
+            nextTimestampsCommand.Parameters.AddWithValue(
+                CursorParameterName,
+                since < SqlDateTime.MinValue ? SqlDateTime.MinValue : since);
+
+            var timestamps = new List<DateTime>();
+            using (var timestampsReader = await nextTimestampsCommand.ExecuteReaderAsync())
+            {
+                while (await timestampsReader.ReadAsync())
+                {
+                    timestamps.Add(timestampsReader.ReadDateTime(columnName).ForceUtc());
+                }
+            }
+
+            return timestamps;
+        }
+
+        public async Task<IReadOnlyCollection<FeedPackageDetails>> GetPackageDetailsForTimestampsAsync(
+            SqlConnection sqlConnection,
+            SqlTransaction sqlTransaction,
+            string columnName,
+            IReadOnlyCollection<DateTime> timestamps)
+        {
+            var packages = new List<FeedPackageDetails>();
+            var packageQuery = $@"
+                        SELECT 
+                            PR.[Id],
+                            P.[NormalizedVersion],
+                            P.[Created],
+                            P.[LastEdited],
+                            P.[Published],
+                            P.[Listed],
+                            P.[HideLicenseReport],
+                            P.[LicenseNames],
+                            P.[LicenseReportUrl]
+                        FROM [dbo].[Packages] AS P
+                        INNER JOIN [dbo].[PackageRegistrations] AS PR ON P.[PackageRegistrationKey] = PR.[Key]
+                        WHERE P.[{columnName}] = @{CursorParameterName}";
+
+            foreach (var timestamp in timestamps)
+            {
+                var packagesThisTimestamp = new List<FeedPackageDetails>();
+
+                var packagesCommand = new SqlCommand(packageQuery, sqlConnection, sqlTransaction)
+                {
+                    CommandTimeout = _commandTimeout
+                };
+
+                packagesCommand.Parameters.AddWithValue("Cursor", timestamp);
+
+                using (var packagesReader = await packagesCommand.ExecuteReaderAsync())
```

Comment:
correct me if I'm wrong, this might execute 20 SQL queries in a single loop (if we read 20 timestamps, and run a query per timestamp. 

Wouldn't it be more performant to create a single query? There are many options to achieve this. One example: get packages in the range between min timestamp and max timestamp. Also, MaxPageSize can be part of the query (you might need to add logic to ignore packages from the last timestamp if there's more then page size).